backend/src/crd/controller/viewer/reconciler/reconciler.go

```diff
@@ -0,0 +1,274 @@
+	if err := r.maybeDeleteOldestViewer(view.Spec.Type); err != nil {
+		// Couldn't delete. Requeue.
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// Set up potential deployment.
+	dpl, err := deploymentFrom(view)
+	if err != nil {
+		utilruntime.HandleError(err)
+		// User error, don't requeue key.
+		return reconcile.Result{}, nil
+	}
+
+	// Set the deployment to be owned by the view instance. This ensures that
+	// deleting the viewer instance will delete the deployment as well.
+	if err := controllerutil.SetControllerReference(view, dpl, r.scheme); err != nil {
+		// Error means that the deployment is already owned by some other instance.
+		utilruntime.HandleError(err)
+		return reconcile.Result{}, err
+	}
+
+	foundDpl := &appsv1.Deployment{}
+	nsn := types.NamespacedName{Name: dpl.Name, Namespace: dpl.Namespace}
+	if err := r.Client.Get(context.Background(), nsn, foundDpl); err != nil {
+		if errors.IsNotFound(err) {
+			// Create a new instance.
+			if err := r.Client.Create(context.Background(), dpl); err != nil {
+				utilruntime.HandleError(fmt.Errorf("error creating deployment: %v", err))
+				return reconcile.Result{}, err
+			}
+		} else {
+			// Some other error.
+			utilruntime.HandleError(err)
+			return reconcile.Result{}, err
+		}
+	}
+	glog.Infof("Created new deployment with spec: %+v", dpl)
+
+	// Set up a service for the deployment above.
+	svc := serviceFrom(view, dpl.Name)
+	// Set the service to be owned by the view instance as well.
+	if err := controllerutil.SetControllerReference(view, svc, r.scheme); err != nil {
+		// Error means that the service is already owned by some other instance.
+		utilruntime.HandleError(err)
+		return reconcile.Result{}, err
+	}
+
+	foundSvc := &corev1.Service{}
+	nsn = types.NamespacedName{Name: svc.Name, Namespace: svc.Namespace}
+	if err := r.Client.Get(context.Background(), nsn, foundSvc); err != nil {
+		if errors.IsNotFound(err) {
+			// Create a new instance.
+			if err := r.Client.Create(context.Background(), svc); err != nil {
+				utilruntime.HandleError(fmt.Errorf("error creating service: %v", err))
+				return reconcile.Result{}, err
+			}
+		} else {
+			// Some other error.
+			utilruntime.HandleError(err)
+			return reconcile.Result{}, err
+		}
+	}
+	glog.Infof("Created new service with spec: %+v", svc)
+
+	return reconcile.Result{}, nil
+}
+
+func setPodSpecForTensorboard(view *viewerV1alpha1.Viewer, s *corev1.PodSpec) {
+	c := corev1.Container{
+		Name:  view.Name + "-pod",
+		Image: "tensorflow/tensorflow:1.11.0",
+		Args: []string{
+			"tensorboard",
+			fmt.Sprintf("--logdir=%s", view.Spec.TensorboardSpec.LogDir),
+			fmt.Sprintf("--path_prefix=/tensorboard/%s/", view.Name),
+		},
+		Ports: []corev1.ContainerPort{
+			corev1.ContainerPort{ContainerPort: viewerTargetPort},
+		},
+	}
+	s.Containers = append(s.Containers, c)
+}
+
+func deploymentFrom(view *viewerV1alpha1.Viewer) (*appsv1.Deployment, error) {
```

Comment:
is it possible to launch a TB pod with a volume mounted, instead of a gcs path?
I guess it's as simple as passing right spec to PodTemplateSpec. is that correct?

Reply:
Pretty much. I did have to make a slight change to how I set up the container so I can reuse the existing volume mount if one is specified. I also added a unit test for this, and verified it works in my own cluster. I will add testing this functionality as part of my e2e test in an upcoming PR.